Collect occurrences information

panglesd opened the pull request:
This PR is not fully finished, but I already open it to already leave the possibility to comment on the approach.

It adds to the source information the occurrences of each identifier.
This information can be used for at least two things:

- Provide a "jump to documentation" from the source rendering (see for example [this file](https://choum.net/panglesd/odoc_occurrences/source/src/xref2/lookup_failures.ml/index.html), which has links to the documentation of `List.fold_left`, ...)
- Count the number of occurrences of an identifier, for instance as a new information to rank search results. For instance, on `odoc` and all of its dependencies, we can find the number of occurrences for each identifier in [this file](https://choum.net/panglesd/odoc_occurrences/occurrences.txt), generated by the reference driver.

Concretely, this PR adds:

- A `--count-occurrences` flag to the `compile` command, which makes `odoc` look for a `.cmt` file and gather all occurrences in it (as odoc `Path`s) as source information. If the source rendering is also activated, this information will be used
- A `count-occurences` command, which Marshal a `Map` from odoc `Identifier.t` to `int`, corresponding to the number of occurrences.

What can be improved:

- [ ] Benchmark this PR in terms of size and time
- ~~Add information of other kinds of nodes, such as constructors...~~ Delayed to future PR
- ~~Handle "local" identifier (they are currently not counted in the total, neither they have links to documentation in the rendered source code)~~ Delayed to future PR
- [x] Handle resolved but hidden identifiers in source rendering (they cannot jump to documentation since there is no generated documentation, but maybe could jump to implementation?)
- [x] testing can be improved

panglesd commented:
I added path to constructors and their resolving function to the PR.
In terms of occurrences, both occurrences of constructors in pattern-matching and in expressions are counted.

I'd particularly appreciate a review on the code related to the addition and resolving of value path and constructor path!

gpetiot commented on src/loader/implementation.ml:
```diff
@@ -503,22 +503,30 @@ let of_cmt (source_id : Odoc_model.Paths.Identifier.SourcePage.t)
+      match (source_id_opt, count_occurrences, cmt_info.cmt_annots) with
+      | Some source_id, _, Implementation impl ->
+          let map, source_infos = of_cmt source_id id impl uid_to_loc in
+          let occ_infos = Occurrences.of_cmt impl in
```
We don't check the value of `count_occurrences` here?

panglesd replied:
No, because when we render the source, we anyway want to collect the occurrences information, to render links from source to documentation!

gpetiot replied:
Alright, makes sense.

panglesd commented on test/occurrences/double_wrapped.t/run.t:
```diff
@@ -0,0 +1,51 @@
+  $ occurrences_print occurrences.txt
+  Main.A.x was used 2 times
+  Main.A.t was used 1 times
+  Main.A was used 3 times
```
Ah, that's a good point, thanks...

I think it shows that the id hashtable is not the best structure: we could use a tree of hashtables instead, counting the direct and indirect occurrences. I'll do that soon!

panglesd replied:
Addressed in 345aa315e549b19bb56ced8b710c0b8c33423b23

panglesd commented on src/xref2/link.ml:
```diff
@@ -351,13 +398,50 @@ and open_ env parent = function
-        let sg = signature env (t.id :> Id.Signature.t) sg in
-        Module sg
-    | Pack _ as p -> p
+    if t.Lang.Compilation_unit.linked || t.hidden then t.content
```
Before, we never really wanted to link hidden modules.
Now, because of occurrence counting, we might want to link hidden modules: actually, with dune's wrapping, there chances that most occurrence will be in a hidden module.

However, this is indeed "extra caution", because when passing the module to link, this [code](https://github.com/ocaml/odoc/blob/master/src/odoc/odoc_link.ml#L11-L16) removes the content of the module if it's hidden!
I would say that the extra caution is not a problem.

gpetiot replied:
Alright, thanks

panglesd commented on src/model/lang.ml:
```diff
@@ -21,12 +21,18 @@ module Source_info = struct
     | Value of Paths.Identifier.SourceLocation.t
+    | ValuePath of Path.Value.t
+    | ModulePath of Path.Module.t
+    | ClassPath of Path.ClassType.t
```
Fixed in 5a411b1.

panglesd commented:
Here is the state of this PR:

### Jump from occurrence to implementation

It works quite well I think! Both inside a file, and "cross module".
The way it works is that ocaml paths are turned into odoc path, which are then turned into shapes, which are resolved.

### Jump from occurrence to documentation

This works well for cross-compilation-unit occurrences, but not intra-compilation unit.
Currently, the ocaml path is turned into an odoc path, which is then resolved to jump to documentation.
Some paths are not rewritten to follow canonical. I'm not sure how to fix that exactly...

The way the "jump to documentation" links are display is temporary. 

### Count occurrences

There is a command to "count occurrences". It creates a "tree of occurrence information", where each identifier has a number of direct and indirect occurrences. A node is an identifier's information, the children are the information of all its "subidentifiers.
This command has options to control what we add in the table: hidden identifiers (or not), and only "persistent" identifiers (occurrences of something in its implementation unit will not be counted), or everything.
The "tree" table is marshalled in a file.

There is also a command to "aggregate" occurrences, which takes several marshalled tables, unmarshall them, and aggregate them in one table which is then marshall.

### Constructors

Constuctors are a bit harder to resolve (there can be several `M.A` which do not resolve to the same constructor depending on the type of the expression), so despite some code being included in this PR, there is no support for constructor occurrences yet.

### The test

A good way of looking at the current state and its problems is by reading the test.

### Conclusion

You can play with the jump to occurrences and implementation [here](https://choum.net/panglesd/odoc/odoc_with_occurrences/html/odoc/source/src/xref2/compile.pp.ml.html).
Please report any wrong behaviour!

panglesd commented:
I have removed the two things that I could not get right in this PR:
- Resolving of "internal occurrences". They were resolved in the ".mli" context, while they belong to the ".ml" context. Now, they are simply not counted, only "persistent" occurrences are counted. Internal occurrence counts can be added back in a future PR.
- Rendering of "jump to documentation" links. They are hard to render in a way that accomodates laptops, touch devices, and which do not mess with the alignment in the code. They can be designed and added in another PR.

Julow commented on src/odoc/occurrences.ml:
```diff
@@ -0,0 +1,203 @@
+    ()
+  in
+  fold_dirs ~dirs:directories ~f ~init:() >>= fun () ->
+  Fs.Directory.mkdir_p (Fs.File.dirname dst);
+  let oc = open_out_bin (Fs.File.to_string dst) in
```
Note for later: It seems that all Odoc commands create the parent directories of files they create. It would be nice to have this written down in the form of a factorized function.

Julow commented on src/odoc/occurrences.ml:
```diff
@@ -0,0 +1,203 @@
+    (Ok []) input
+  >>= fun files -> Ok (List.concat files)
+
+let aggregate files file_list ~warnings_options:_ ~dst =
```
It seems that the input and output files could be large. What do you think of using a sorted line-based format for both the inputs and the output ?
Sorted files can be merged in a streaming way and the command no longer needs to hold all the inputs and outputs in memory at the same time.

panglesd replied:
I fully agree that a data structure that would allow to merge several "occurrence tables" (even many of them) in `O(n)` time (`n` being the total number of elements) and `O(t)` space (`t` being the number of tables to merge) is possible as you suggest, interesting and a good thing to have, let's do that in a future PR!

Julow commented on src/xref2/compile.ml:
```diff
@@ -90,7 +90,28 @@ let rec unit env t =
+            | ModuleType v -> ModuleType (map_doc (module_type_path env) v)
+            | Type v -> Type (map_doc (type_path env) v)
+            | Constructor v -> Constructor (map_doc (constructor_path env) v)
+            | i -> i
```
What about class types ? I suggest writing a exhaustive match to avoid surprise in the future.

panglesd commented on doc/driver.mld:
```diff
@@ -750,6 +760,7 @@ let compiled = compile_all () in
 let linked = link_all compiled in
 let () = index_generate () in
 let _ = js_index () in
+let _ = count_occurrences (Fpath.v "occurrences.txt") in
```
The "print_occurrences" binary defined in the [test](https://github.com/ocaml/odoc/pull/976/files#diff-c6fcd61e392e9c4948f3ee6cb0cab96d03d34ba78a5da606ae5921fd7840a231) allows to render it in a readable way. Do you think I should include in `odoc` itself a way to render it as readable?

Also, maybe the `count-occurrences` command should produce files with a "fixed" extension, e.g. `*.occ-odoc`.

panglesd replied:
I took inspiration from the constraints we give to the output name of a file representing a source tree.
So, the output file name of the `count-occurrences` and `aggregate-occurrences` must now have `occurrences-` prefix and `odoc` extension.

panglesd commented on src/model/lang.ml:
```diff
@@ -18,15 +18,31 @@ open Paths
+    | Unresolved of 'a
+    | Resolved of Identifier.SourceLocation.t
+
+  type ('doc, 'impl) jump_to = {
```
No... (the typedtree traverse used to output a list of `jump_to` where `'doc` and `'impl` were different, but that's not anymore the case, good catch!)

panglesd commented on src/model/paths.ml:
```diff
@@ -19,6 +19,15 @@ module Ocaml_env = Env
 
 open Names
 
+let contains_double_underscore s =
```
I exposed it in `Names` and every other module uses that. (It was exposed in `Root`, so it's not anymore)

panglesd commented on src/model/paths.ml:
```diff
@@ -668,7 +712,7 @@ module Path = struct
         when Names.ModuleName.is_internal m ->
           true
-      | `Identifier _ -> false
+      | `Identifier i -> Identifier.is_internal_rec i
```
Recursive check removed. It was used to not include occurrences such as `Main__A.x`, which came up when counting "non-persistent" occurrences...which we do not include anymore. Anyway, those should have been resolved to `Main.A.x`. We'll solve that later, in the PR which includes non-persistent occurrences.

panglesd commented on src/loader/typedtree_traverse.ml:
```diff
@@ -45,19 +55,85 @@ module Analysis = struct
+    | { Typedtree.ctyp_desc = Ttyp_constr (p, _, _); ctyp_loc; _ }
+      when not ctyp_loc.loc_ghost ->
+        poses := (Type p, ctyp_loc) :: !poses
+    | _ -> ()
```
Good catch! In fact, we only count occurrences of class types. It is less useful than counting occurrences of classes, but the infrastructure to resolve class types was already there.
Counting classes can be added in a future PR.

panglesd commented on test/occurrences/double_wrapped.t/run.t:
```diff
@@ -0,0 +1,137 @@
+  Main.A.x was used directly 2 times and indirectly 0 times
+  Main.B was used directly 1 times and indirectly 0 times
+  Main__ was used directly 0 times and indirectly 2 times
+  Main__.C was used directly 1 times and indirectly 1 times
```
Good question...

The specificity of `Main__.C` (compared to `Main__.A` and `Main__.B`) is that it has a canonical which does not exists.

panglesd replied:
Ah, I think that aliases are "auto-canonical" when one hidden and the other not. In this case, both are hidden, so neither is a canonical of the other.

So, the problem is not about canonicals, but about aliases, as you also mention in the next comment. Should occurrences of aliases be counted in a single bucket? We can discuss that :)

Julow replied:
A broken count for `Main__.C` makes sense as it's build in a broken way.

panglesd commented on src/xref2/link.ml:
```diff
@@ -400,16 +415,180 @@ and open_ env parent = function
+      | Pack _ as p -> p
+  in
+  let source_info =
+    let env = Build_env.unit env t in
```
Good catch, this was needed when counting "internal occurrences". They needed the entire environment built to resolve. However, they still weren't resolved against the right environment: they needed the environment of the _implementation_, but were given environment of the interface.

Since we only count persistent occurrences, I removed all this module.

panglesd commented on src/xref2/link.ml:
```diff
@@ -400,16 +415,180 @@ and open_ env parent = function
+                    Value
+                      (jump_to v
+                         (Shape_tools.lookup_value_path env)
+                         (value_path ~report_errors:false env))
```
In the end, there are much less failures than before (that's again because we removed the non-persistent identifiers!).

So, I'm in favor of considering such failures as normal failures, and printing them if the options say so.

(With non-persistent occurrences, there were warnings that one could never get rid of, which is why I introduced the non-error reporting)

panglesd commented on src/xref2/link.ml:
```diff
@@ -1087,8 +1266,7 @@ and type_expression : Env.t -> Id.Signature.t -> _ -> _ =
 let link ~filename x y =
-  Lookup_failures.catch_failures ~filename (fun () ->
-      if y.Lang.Compilation_unit.linked || y.hidden then y else unit x y)
+  Lookup_failures.catch_failures ~filename (fun () -> unit x y)
```
You are right, fixed.

panglesd commented on test/occurrences/double_wrapped.t/run.t:
```diff
@@ -0,0 +1,137 @@
+  Main__ was used directly 0 times and indirectly 2 times
+  Main__.C was used directly 1 times and indirectly 1 times
+  Main__.C.y was used directly 1 times and indirectly 0 times
+  Main__A was used directly 1 times and indirectly 0 times
```
`Main__A` is a hidden module.
An alias `Main__.A` is defined in `main__.ml`:
```
module A = Main__A
(** @canonical Main.A *)
```
so, `Main__.A` has `Main.A` as canonical.

However, I don't think that `Main__A` has `Main.A` as canonical: it is aliased with something _which is itself also hidden_. So, it is not "auto-canonicalised", and the `@canonical` tag does not affect it.

One way to fix that would be to say that when we have
```
module X = Y
(** @canonical C *)
```
_and_ `Y` is hidden, then both `X` and `Y` have `C` as canonical. Do you agree? Do you think I should do that in this PR?

Julow replied:
> and Y is hidden, then both X and Y have C as canonical. Do you agree?

That might already be the case ? The test in `test/xref2/canonical_nested.t/` seems show the same pattern. Perhaps you might need to specially handle `` `Canonical `` identifiers when collecting occurrences ?

> Do you think I should do that in this PR?

No but it would be nice to have a clear picture of the problem.

panglesd replied:
> That might already be the case ?

I tried and I don't think so. Here is my test:

```ocaml
module Cano = struct
  type t
end

module Hidden__one = struct
  type t
end

module Hidden__two = Hidden__one
(** @canonical Test_odoc.A.Cano *)

let y : Hidden__one.t = assert false
let z : Hidden__two.t = assert false
```

In the rendered doc, `z` is shown as having type `Cano.t`, while `y` is shown as having type `{Hidden__one}1.t`.

Julow commented on src/xref2/link.ml:
```diff
@@ -400,16 +415,180 @@ and open_ env parent = function
+      (fun env item ->
+        match item with
+        | Module (_, m) -> module_ env m
+        | ModuleSubstitution m -> Env.open_module_substitution m env
```
This code is no longer there :)

Julow commented on src/xref2/tools.ml:
```diff
@@ -878,6 +878,29 @@ and lookup_type_gpath :
+    | `FValue (name, t) :: _ -> Ok (`FValue (name, Subst.value sub t))
+    | [] -> Error `Find_failure
+  in
+  let res =
```
It's pre-indented in case we want to add memoisation ? :p

Julow commented on src/xref2/link.ml:
```diff
@@ -403,13 +400,74 @@ and open_ env parent = function
+    | Some inf ->
+        let jump_to v f_impl f_doc =
+          let documentation =
+            match v.documentation with Some p -> Some (f_doc p) | None -> None
```
As documentation link are not used, would it make sense to comment out this code ? It would be a shame to loose it though.

panglesd replied:
"Links to documentation" are used to count occurrences, I propose to keep that feature, otherwise I need to rename the PR ^^.
The name is wrong though. Maybe `type jump_to = {documentation : ... ; implementation : ...}` could be renamed to `type occurrence_of = ...`? (and this function would also be renamed)

Julow replied:
Right! I focused on the html generator and noticed that it wasn't used there and wrongly assumed it wasn't used at all.
Perhaps the code in `document` could be commented-out ?

panglesd replied:
I commented the code in document extracting an href to documentation from an occurrence.

Julow commented on test/occurrences/double_wrapped.t/b.ml:
```diff
@@ -0,0 +1,15 @@
+module Y = A
```
Missing a test for constructors and class types. Also, there could be a type alias and a value alias.

panglesd replied:
I completely removed support for constructors.
It was broken as it relied on a broken concept of constructor paths. Constructors are resolved differently: they can be several constructors with the same path, they are disambiguated using types. This was not done here. I think support for constructors should be in another PR.

I added a test for class types, but I think support for class types could be removed as well, for being too incomplete. It is I think useless to count class types when nothing is done about classes.

Julow replied:
It sounds reasonable to remove support for constructors and class types.

panglesd replied:
Done.

Julow commented on src/odoc/occurrences.ml:
```diff
@@ -0,0 +1,205 @@
+    let rec add ?(kind = `Indirect) id =
+      let incr htbl id =
+        let { direct; indirect; sub } =
+          match H.find_opt htbl id with Some n -> n | None -> v_item ()
```
Compat with 4.02:
```suggestion
          try H.find htbl id with Not_found -> v_item ()
```

panglesd replied:
I'll wait for #1060 for those kind of issues.

Julow commented on src/odoc/occurrences.ml:
```diff
@@ -0,0 +1,205 @@
+
+let handle_file file ~f =
+  Odoc_file.load file
+  |> Result.map @@ fun unit' ->
```
That's not compatible.

panglesd replied:
CI i now passing, compatibility should be achieved...

jonludlam commented:
Could we please avoid the change from Map to Hashtbl in Ident_env? The reasoning behind this is a change suggested by @lpw25 a while back to resolve the 'root' of a reference during the load phase so that references end up looking more like paths. The reasoning behind _that_ is to have lexical scoping of references rather than the dynamic scoping we currently have - that way we can avoid having to make [this sort of change](https://github.com/ocaml/ocaml/pull/10815/files#diff-9bc0444e1c79002ed50d614ffbd02ffe205d1cab05c50beac9620cc8ef90a7cdL242-R242). Switching to Hashtbls will make that much trickier. 

This is somewhat contingent on my guess that the change to ident_env is separable from the rest of the PR - if it isn't we might have to just commit this and solve the reference issue when we get around to it.

panglesd commented:
> Could we please avoid the change from Map to Hashtbl in Ident_env?

Sure, I'll revert that. Again, it is part of the changes I made for non-persistent occurrences, that are not included anymore in this PR. I left it as it seemed an improvement, but that was not deeply though!

Julow reviewed (approved):
This is now perfect :) Let's merge!

jonludlam commented:
I agree, lovely stuff! We might want to revisit how and when we process stuff (see https://hackmd.io/__sMo7xRSEGDephPhtDpNQ for some thoughts), but this is all consistent with the current handling so no sense in delaying it.
